### ISSUE TITLE
Structured-Self-Attentive-Sentence-Embedding

### ISSUE THREAD

**Issue by rangwani-harsh** [allenai/allennlp#2188]
Hi team,
I have written an implementation for Structured-Self-Attentive-Sentence-Embedding https://arxiv.org/pdf/1703.03130.pdf which is basically a Seq2Vec method. I was not sure of how good it would be to include that as a part of Allennlp but now after [pytext](https://github.com/facebookresearch/pytext) have released it as a part of their library. I think it would be better to have it as a part of allennlp too although I don't know that if there already exists something in allennlp that does a similar thing with better performance already present. 

In case we feel good about it I can probably rewrite the implementation [here](https://github.com/rangwani-harsh/irony-detection-research/blob/3aa4ae1f87bd646f9c556c5d793561e72697140f/irony_model/module/modules.py#L143) as per the AllenNLP style guide

**Comment by matt-gardner** [comment]
Seems like a reasonable thing to include to me.  If you're willing to match our style guide and write tests, we'd be happy to have the contribution!

**Comment by rangwani-harsh** [comment]
There are two essential things that I require on clarity before raising the PR.

1) As a Seq2Vec encoder, the input of shape [batch_size, time_steps, hidden_dims] needs to be converted to the output of shape [batch_size, dims].  If we use multiple attention_heads we will finally have the representation of the shape [batch_size, attention_heads, hidden_dims]. So there are essentially two methods:
- Take the average of all the attention_heads and output the tensor of the shape [batch_size, hidden_dims]
- Reshape the output to be of the shape [batch_size, attention_heads*hidden_dims]. (Originally implemented in the paper) 
2) The other is that in this paper for ensuring that all attention heads attend to different portions of the text there is a forbenius norm based penalty. As an output of the forward function I need to atleast provide the representation along with the penalty loss term. So my output needs to be a dictionary.
Which is unlike the other seq2vec encoders already present.

**Comment by matt-gardner** [comment]
1. Either option seems reasonable; I'd default to the second one you suggested, though you could have an option for either if you feel so inclined.
2. Yes, you're right that returning an additional penalty breaks the Seq2Vec API.  I can think of two ways to handle this:
- Have a flag that enables the computation (and return) of this penalty, and only return a dictionary in that case.  This lets you use the module as a `Seq2VecEncoder` if you pass it the right flags, and it lets you use it with the additional penalty if you want (where now it is no longer strictly a `Seq2VecEncoder`).  We have a few other modules that behave this way, I believe.
- Allow `forward` to take a mutable object (like a list) that you append the loss to if the argument is present.  This also breaks the `Seq2VecEncoder` API a little bit, but not as much, and we could consider adding `**kwargs` as an argument to the base class.  I'm not certain how I feel about this option, but it's an option.

If anyone else has opinions about these two options, feel free to weigh in.

**Comment by rangwani-harsh** [comment]
In response to the second method that you have mentioned, I think if we will be using any attention based encoders in the neural network pipeline there might be a use case that for visualizing attention we might have to return attention weights. So having a generic way of passing any additional data that may be required can be done in this way. If we think this is a good way I can go about to thinking a way in which we would be able to implement it.

**Comment by joelgrus** [comment]
I think my preferred solution would actually be something like the following:

(1) a Structured-Self-Attentive-Sentence-Embedding `Module` that returns a dict with the embedding and the loss function and whatever else
(2) a separate Structured-Self-Attentive-Sentence-Embedding `Seq2VecEncoder` that just wraps the module but respects the `Seq2VecEncoder` API

that way you have to be pretty explicit about how you use it and you're not breaking any APIs.

if this is something we thing we'd need to do a lot, we could even create a `Seq2VecModuleWrapper` or something that just takes a module and a key and creates the corresponding Seq2VecEncoder

what do you think about this solution?

**Comment by matt-gardner** [comment]
That sounds totally reasonable to me.

**Comment by rangwani-harsh** [comment]
Hi team,
I needed some suggestions before I implement the wrapper. As we wont be able to use the forbenius regularization loss in the Seq2Vec encoder. So at present I am not sure how much will be the effect of implementing multi headed attention will be there. As we wont be able to use regularization to penalise attending of heads on the same parts of sentence. So there are now potentially two ways in which I could possibly implement the wrapper.

(1) Don't allow multihead attention in the Seq2Vec wrapper (similar to pytext)
(2) Allow multiheaded attention without regularization term.

Thanks
